deps/key-value-storage/src/postgres/mod.rs

```diff
@@ -0,0 +1,240 @@
+    #[derivative(Debug = "ignore")]
+    #[serde(default)]
+    pub password: Option<String>,
+
+    /// The port of the PostgreSQL database.
+    #[serde(default = "default_port")]
+    pub port: u16,
+
+    #[serde(default = "default_host")]
+    pub host: String,
+
+    /// The name of the policy table.
+    #[serde(default = "default_table")]
+    pub table: String,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            db: default_db(),
+            username: default_username(),
+            password: None,
+            port: default_port(),
+            host: default_host(),
+            table: default_table(),
+        }
+    }
+}
+
+fn default_db() -> String {
+    "postgres".to_string()
+}
+
+fn default_username() -> String {
+    "postgres".to_string()
+}
+
+fn default_host() -> String {
+    "localhost".to_string()
+}
+
+fn default_port() -> u16 {
+    5432
+}
+
+fn default_table() -> String {
+    "key_value".to_string()
+}
+
+pub struct PostgresClient {
+    pool: Arc<PgPool>,
+    table: String,
+}
+
+impl PostgresClient {
+    pub async fn new(config: Config) -> Result<Self> {
+        info!("Initializing PostgreSQL client");
+        debug!("Connecting to PostgreSQL DB: {config:?}");
+        let auth_part = config
+            .password
+            .map(|password| format!("{}:{password}@", config.username))
+            .unwrap_or(format!("{}@", config.username));
+
+        let url = format!(
+            "postgres://{}{}:{}/{}",
+            auth_part, config.host, config.port, config.db
+        );
+
+        let pool = PgPoolOptions::new()
+            .max_connections(5)
+            .connect(&url)
+            .await
+            .context("failed to connect to PostgreSQL DB")
+            .map_err(|e| KeyValueStorageError::InitializeBackendFailed {
+                source: anyhow!("failed to connect to PostgreSQL DB: {e}"),
+            })?;
+
+        Ok(Self {
+            pool: Arc::new(pool),
+            table: config.table,
+        })
+    }
+
+    /// Check if the key is valid.
+    ///
+    /// The key is valid if it only contains ASCII alphanumeric characters, `-` or `_`.
+    /// No spaces and other special characters are allowed to prevent SQL injection.
+    fn is_valid_key(key: &str) -> bool {
+        key.chars()
+            .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_')
+    }
+}
+
+pub struct PolicyItem {
+    pub key: String,
+    pub value: String,
+}
+
+#[async_trait]
+impl KeyValueStorage for PostgresClient {
+    #[instrument(skip_all, name = "PostgresClient::set_key")]
+    async fn set_key(&self, key: String, value: String, overwrite: bool) -> Result<()> {
+        if !Self::is_valid_key(&key) {
+            return Err(KeyValueStorageError::SetKeyFailed {
+                source: anyhow::anyhow!("key contains invalid characters"),
+                key,
+            });
+        }
+
+        // we do the base64 encoding here to avoid SQL injection.
```

Comment:
I don't need think we need to convert the value payload to base64 to avoid injections, since the sqlx queries are parameterized statements `value` is bound. Maybe the column-type for value should be a byte array? (`bytea` in postgres schema)